code-experiments/build/python/example_experiment_non_anytime.py

```diff
@@ -0,0 +1,165 @@
+    return unique(logspace(0, log10(max_budget), num=num).astype(int))
+
+
+def fmin(problem, x0, solver, budget):
+    """Invokes `solver` on `problem` with initial solution `x0`.
+
+    `solver` should evaluate the final/returned solution. Add here the interface to your solver.
+    """
+    if solver.__name__ == 'fmin' and solver.__globals__['__name__'] in ['cma', 'cma.evolution_strategy', 'cma.es']:
+        return solver(problem, x0, 2, {'verbose': -9})  # set 'verbose': 1 for more output
+
+    elif solver.__name__ == 'fmin' and solver.__globals__['__name__'] == 'scipy.optimize.optimize':
+        return solver(problem, x0, xtol=1e-9, ftol=1e-9, maxfun=budget, disp=False)  # set disp=True for more output
+
+    elif solver.__name__ == 'differential_evolution':
+        pop_size_multiplier = 5  # pop_size = pop_size_multiplier * problem.dimension
+        # a strange but correct way to set max_iter so that the budget is respected
+        max_iter = max(1, int(budget / pop_size_multiplier / problem.dimension) - 1)
+        bounds = [x for x in zip(problem.lower_bounds, problem.upper_bounds)]
+        return solver(problem, bounds, popsize=pop_size_multiplier, maxiter=max_iter, strategy='best1bin',
+                      tol=1e-9, polish=False, disp=False)  # set disp=True for more output
+
+    else:
+        return solver(problem, x0)
+
+
+def main():
+    ### input
+    suite_name = "bbob"
+    solver = scipy.optimize.fmin  # the Nelder-Mead downhill simplex method
+    # solver = scipy.optimize.differential_evolution  # to use differential evolution instead
+    # solver = cma.fmin  # to use CMA-ES instead
+    algorithm_name = "Nelder-Mead"  # no spaces allowed
+    output_folder = algorithm_name  # no spaces allowed
+
+    # a list of increasing budgets (to be additionally multiplied by dimension)
+    # gradually increase `max_budget` to 10, 1000, ... or replace with a user-defined list
+    budget_multiplier_list = default_budget_list(max_budget=5)
+
+    ### prepare (see http://numbbo.github.io/coco-doc/C/ for suite and observer parameters)
+    suite = cocoex.Suite(suite_name, "", "")
+    observer = cocoex.Observer(suite_name, "result_folder: {} algorithm_name: {}".format(output_folder, algorithm_name))
+
+    ### go
+    for index, _ in enumerate(suite):
+        with ProblemNonAnytime(suite, observer, index) as problem:
+            problem.print_progress()
+            for budget in [problem.dimension * x for x in budget_multiplier_list]:  # iterate over budgets
+                problem.reset()
+                x0 = problem.initial_solution  # initial solution
+                # apply restarts while neither the problem is solved nor the budget is exhausted
+                while problem.evaluations < budget and not problem.final_target_hit:
+                    fmin(problem, x0, solver, budget)
+                    x0 = problem.random_solution  # a random initial solution for restarted algorithms
+            problem.free()
```

Comment:
This line seems superfluous, because here problem.__exit__() is called anyways. That is kind-of the point of the `with` construct.